File: frontend/src/api/aicalling.ts
import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import { api } from "./client";

/* ---------- Тест TTS ---------- */

export interface TTSRole {
  id: string;
  label: string;
}
export interface TTSVoice {
  id: string;
  roles: TTSRole[];
}

export function useTtsVoices() {
  return useQuery<TTSVoice[]>({
    queryKey: ["ai-calling", "voices"],
    queryFn: async () => {
      const { data } = await api.get("/ai-calling/voices");
      return data.voices ?? [];
    },
    staleTime: 5 * 60 * 1000,
  });
}

export interface TtsTestRequest {
  text: string;
  voice?: string;
  speed?: number;
  role?: string;
}
export interface TtsTestResponse {
  audio_base64: string;
  format: string;
  sample_rate: number;
  size_bytes: number;
}

export function useTtsTest() {
  return useMutation<TtsTestResponse, Error, TtsTestRequest>({
    mutationFn: async (body) => {
      const { data } = await api.post("/ai-calling/tts-test", body);
      return data;
    },
  });
}

/* ---------- Тест диалога (v2.0) ---------- */

export interface DialogTurnResponse {
  robot_text: string;
  phase: string;
  phase_label: string;
  node: string;
  qual_step: number;
}

export function useDialogStart() {
  return useMutation<DialogTurnResponse, Error, { session_id: string }>({
    mutationFn: async (body) => {
      const { data } = await api.post("/ai-calling/dialog/start", body);
      return data;
    },
  });
}

export function useDialogTurn() {
  return useMutation<DialogTurnResponse, Error, { session_id: string; user_text: string }>({
    mutationFn: async (body) => {
      const { data } = await api.post("/ai-calling/dialog/turn", body);
      return data;
    },
  });
}

export function useDialogDeleteSession() {
  return useMutation<unknown, Error, string>({
    mutationFn: async (sessionId) => {
      const { data } = await api.delete(`/ai-calling/dialog/session/${sessionId}`);
      return data;
    },
  });
}

/* ---------- Скрипты диалога ---------- */

export interface Scenario {
  id: string;
  name: string;
  description?: string;
}

export function useScenarios() {
  return useQuery<Scenario[]>({
    queryKey: ["ai-calling", "scenarios"],
    queryFn: async () => {
      const { data } = await api.get("/ai-calling/scenarios");
      return data.scenarios ?? [];
    },
    staleTime: 60 * 1000,
  });
}

export interface ScriptCorrection {
  id: string;
  trigger: string;
  current_answer?: string;
  correct_answer: string;
  phase: string;
  enabled: boolean;
}

export function useScriptCorrections() {
  return useQuery<ScriptCorrection[]>({
    queryKey: ["ai-calling", "script-corrections"],
    queryFn: async () => {
      const { data } = await api.get("/ai-calling/script-corrections");
      return data.corrections ?? [];
    },
    staleTime: 30 * 1000,
  });
}

export function useAddScriptCorrection() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (body: Omit<ScriptCorrection, "id">) => {
      const { data } = await api.post("/ai-calling/script-corrections", body);
      return data;
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ["ai-calling", "script-corrections"] }),
  });
}

export function useDeleteScriptCorrection() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      const { data } = await api.delete(`/ai-calling/script-corrections/${id}`);
      return data;
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ["ai-calling", "script-corrections"] }),
  });
}

/* ---------- Кампании обзвона ---------- */

export interface Campaign {
  id: string;
  name: string;
  segment_key: string;
  scenario_id: string;
  status:
    | "scheduled"
    | "running"
    | "waiting_window"
    | "paused"
    | "completed"
    | "cancelled"
    | "failed";
  max_concurrent: number;
  scheduled_at: string | null;
  window_start: string | null;
  window_end: string | null;
  timezone: string;
  total: number;
  completed: number;
  succeeded: number;
  failed: number;
  progress: number;
  started_at: string | null;
  ended_at: string | null;
  error: string | null;
  created_at: string | null;
}

export interface CampaignItem {
  id: string;
  patient_id: string | null;
  phone: string;
  status: "pending" | "calling" | "done" | "no_answer" | "failed" | "cancelled";
  outcome: string | null;
  summary: string | null;
  duration_sec: number | null;
  attempts: number;
  updated_at: string | null;
}

export interface CampaignCreateRequest {
  name: string;
  segment_key: string;
  scenario_id?: string;
  max_concurrent?: number;
  scheduled_at?: string | null;
  window_start?: string | null;
  window_end?: string | null;
  timezone?: string;
}

function campaignsActive(items: Campaign[]): boolean {
  return items.some((c) =>
    ["scheduled", "running", "waiting_window"].includes(c.status),
  );
}

export function useCampaigns() {
  return useQuery<{ items: Campaign[] }>({
    queryKey: ["ai-calling", "campaigns"],
    queryFn: async () => {
      const { data } = await api.get("/ai-calling/campaigns");
      return data;
    },
    // Адаптивный polling: чаще, пока есть активные кампании.
    refetchInterval: (query) =>
      campaignsActive(query.state.data?.items ?? []) ? 3000 : false,
  });
}

export function useCampaignItems(campaignId: string | null) {
  return useQuery<{ items: CampaignItem[] }>({
    queryKey: ["ai-calling", "campaign-items", campaignId],
    queryFn: async () => {
      const { data } = await api.get(`/ai-calling/campaigns/${campaignId}/items`);
      return data;
    },
    enabled: !!campaignId,
    refetchInterval: 4000,
  });
}

export function useCreateCampaign() {
  const qc = useQueryClient();
  return useMutation<Campaign, Error, CampaignCreateRequest>({
    mutationFn: async (body) => {
      const { data } = await api.post("/ai-calling/campaigns", body);
      return data;
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ["ai-calling", "campaigns"] }),
  });
}

export function useCampaignControl() {
  const qc = useQueryClient();
  return useMutation<Campaign, Error, { id: string; action: "start" | "pause" | "resume" | "cancel" }>({
    mutationFn: async ({ id, action }) => {
      const { data } = await api.post(`/ai-calling/campaigns/${id}/control`, { action });
      return data;
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ["ai-calling", "campaigns"] }),
  });
}

/* ---------- Тестовый звонок на телефон ---------- */

export interface TestCallResponse {
  call_id: string;
  status: string;
  greeting?: string;
}

export function useTestCall() {
  return useMutation<TestCallResponse, Error, { phone: string; scenario_id?: string }>({
    mutationFn: async (body) => {
      const { data } = await api.post("/ai-calling/test-call", body);
      return data;
    },
  });
}

export interface CallTranscriptLine {
  role: string; // robot | client | system
  text: string;
  timestamp?: number;
}
export interface CallStatus {
  call_id: string;
  status: string; // active | ringing | completed | failed ...
  transcript: CallTranscriptLine[];
  client_status?: string;
  summary?: string;
  duration?: number | null;
}

export function useCallStatus(callId: string | null) {
  return useQuery<CallStatus>({
    queryKey: ["ai-calling", "call-status", callId],
    queryFn: async () => {
      const { data } = await api.get(`/ai-calling/calls/${callId}`);
      return data;
    },
    enabled: !!callId,
    // Поллим, пока звонок не завершится.
    refetchInterval: (query) => {
      const st = query.state.data?.status;
      return st && ["completed", "failed"].includes(st) ? false : 1500;
    },
  });
}

File: backend/app/models/ai_calling.py
"""ИИ-обзвон: кампании и их элементы (по одному звонку на пациента).

Кампания обзвона = сегмент пациентов + сценарий + расписание/окна. Оркестратор
(Celery) ставит звонки через Asterisk (AMI) и заворачивает их в aicallrobot.
"""

import uuid
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql import func

from app.database import Base


class AiCallingCampaign(Base):
    __tablename__ = "ai_calling_campaigns"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    # Аудитория — сохранённый сегмент пациентов.
    segment_key: Mapped[str] = mapped_column(String(50), nullable=False)
    scenario_id: Mapped[str] = mapped_column(String(100), default="default")
    status: Mapped[str] = mapped_column(
        String(20), default="scheduled", index=True
    )  # scheduled|running|waiting_window|paused|completed|cancelled|failed
    # Сколько звонков вести одновременно (≤ лимита каналов тарифа Novofon).
    max_concurrent: Mapped[int] = mapped_column(Integer, default=1)
    # Расписание и окна обзвона.
    scheduled_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    window_start: Mapped[str | None] = mapped_column(String(5), nullable=True)  # "09:00"
    window_end: Mapped[str | None] = mapped_column(String(5), nullable=True)    # "20:00"
    timezone: Mapped[str] = mapped_column(String(64), default="Europe/Moscow")
    # Голос TTS для звонков кампании (голос/амплуа/скорость).
    tts_voice: Mapped[str | None] = mapped_column(String(50), nullable=True)
    tts_role: Mapped[str | None] = mapped_column(String(50), nullable=True)
    tts_speed: Mapped[float | None] = mapped_column(Float, nullable=True)
    # Счётчики прогресса.
    total: Mapped[int] = mapped_column(Integer, default=0)
    completed: Mapped[int] = mapped_column(Integer, default=0)
    succeeded: Mapped[int] = mapped_column(Integer, default=0)
    failed: Mapped[int] = mapped_column(Integer, default=0)
    created_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    started_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    ended_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    error: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )


class AiCallingCampaignItem(Base):
    __tablename__ = "ai_calling_campaign_items"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    campaign_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("ai_calling_campaigns.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    patient_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("patients.id"), nullable=True, index=True
    )
    phone: Mapped[str] = mapped_column(String(50), nullable=False)
    # call_id, выданный aicallrobot (он же UUID для AudioSocket).
    call_id: Mapped[str | None] = mapped_column(String(64), nullable=True)
    status: Mapped[str] = mapped_column(
        String(20), default="pending", index=True
    )  # pending|calling|done|no_answer|failed|cancelled
    outcome: Mapped[str | None] = mapped_column(String(30), nullable=True)  # client_status
    summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    duration_sec: Mapped[int | None] = mapped_column(Integer, nullable=True)
    comm_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("communications.id"), nullable=True
    )
    attempts: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )
